tests/test_datastructures.py

```diff
@@ -138,6 +145,19 @@ def test_url_from_scope() -> None:
     assert u == "https://example.org/path/to/somewhere?abc=123"
     assert repr(u) == "URL('https://example.org/path/to/somewhere?abc=123')"
 
+    scope_with_host_header = {
+        "scheme": "http",
+        "path": "/some/path",
+        "query_string": b"query=string",
+        "headers": [
+            (b"content-type", b"text/html"),
+            (b"host", b"example.com:8000"),
+            (b"accept", b"text/html"),
+        ],
+    }
+    url_from_scope_with_host = URL(scope=scope_with_host_header)
+    assert url_from_scope_with_host.hostname == "example.com"
+
```

Comment:
I meant doing it here... 
https://github.com/encode/starlette/blob/74ccb961b3c1b1871aa7ed70a81dd3000e0194da/starlette/datastructures.py#L39-L40